File: web/src/main/java/hu/bsstudio/robonaut/skill/SkillGateHandler.java
package hu.bsstudio.robonaut.skill;

import hu.bsstudio.robonaut.race.skill.SkillRaceService;
import hu.bsstudio.robonaut.race.skill.model.GateInformation;
import hu.bsstudio.robonaut.team.model.DetailedTeam;
import lombok.RequiredArgsConstructor;
import org.springframework.web.reactive.function.server.HandlerFunction;
import org.springframework.web.reactive.function.server.ServerRequest;
import org.springframework.web.reactive.function.server.ServerResponse;
import reactor.core.publisher.Mono;

@RequiredArgsConstructor
public class SkillGateHandler implements HandlerFunction<ServerResponse> {

    private final SkillRaceService service;

    @Override
    public Mono<ServerResponse> handle(final ServerRequest request) {
        final var detailedTeam = request.bodyToMono(GateInformation.class)
            .flatMap(service::updateSkillRaceResultOnGate);
        return ServerResponse.ok().body(detailedTeam, DetailedTeam.class);
    }
}

File: web/src/main/kotlin/hu/bsstudio/robonaut/skill/SkillGateHandler.kt
package hu.bsstudio.robonaut.skill

import hu.bsstudio.robonaut.race.skill.SkillRaceService
import hu.bsstudio.robonaut.race.skill.model.GateInformation
import org.springframework.web.reactive.function.server.*
import reactor.core.publisher.Mono

class SkillGateHandler(private val service: SkillRaceService) : HandlerFunction<ServerResponse> {

    override fun handle(request: ServerRequest): Mono<ServerResponse> {
        return request.bodyToMono<GateInformation>()
            .flatMap(service::updateSkillRaceResultOnGate)
            .let { detailedTeam -> ServerResponse.ok().body(detailedTeam) }
    }
}
